src/libraries/Native/Unix/System.Globalization.Native/pal_timeZoneInfo.c

```diff
@@ -80,3 +63,238 @@ int32_t GlobalizationNative_IanaIdToWindowsId(const UChar* ianaId, UChar* window
+    // The reason we get the standard name is because TR35 (LDML) defines a rule that
+    // states that metazone generic names should use standard names if there is no DST
+    // transition within a +/- 184 day range near the timestamp being translated.
+    //
+    // See the "Type Fallback" section in:
+    // https://www.unicode.org/reports/tr35/tr35-dates.html#Using_Time_Zone_Names
+    //
+    // This might make sense when attached to an exact timestamp, but doesn't work well
+    // when using the generic name to pick a time zone from a list.
+    // Note that this test only happens when the generic name comes from a metazone.
+    //
+    // ICU implements this test in TZGNCore::formatGenericNonLocationName in
+    // https://github.com/unicode-org/icu/blob/master/icu4c/source/i18n/tzgnames.cpp
+    // (Note the kDstCheckRange 184-day constant.)
+    //
+    // The rest of the code below is a workaround for this issue.  When the generic
+    // name and standard name match, we search through the other time zones for one
+    // having the same base offset and standard name but a shorter generic name.
+    // That will at least keep them grouped together, though note that if there aren't
+    // any found that means all of them are using the standard name.
+    //
+    // If ICU ever adds an API to get a generic name that doesn't perform the
+    // 184-day check on metazone names, then test for the existence of that new API
+    // and use that instead of this workaround.  Keep the workaround for when the
+    // new API is not available.
+
+    // Get the standard name for this time zone.  (-1 == timeZoneId is null terminated)
+    // Note that we leave the calendar open and close it later so we can also get the base offset.
+    UChar standardName[DISPLAY_NAME_LENGTH];
+    UCalendar* calendar = ucal_open(timeZoneId, -1, locale, UCAL_DEFAULT, err);
+    if (U_FAILURE(*err)) goto exit;
+
+    ucal_setMillis(calendar, timestamp, err);
+    if (U_FAILURE(*err)) goto exit;
+
+    ucal_getTimeZoneDisplayName(calendar, UCAL_STANDARD, locale, standardName, DISPLAY_NAME_LENGTH, err);
+    if (U_FAILURE(*err)) goto exit;
+
+    // See if the generic name is the same as the standard name.
+    if (u_strcmp(genericName, standardName) == 0)
+    {
+        // Get some details for later comparison.
+        const int32_t originalGenericNameActualLength = u_strlen(genericName);
+        const int32_t baseOffset = ucal_get(calendar, UCAL_ZONE_OFFSET, err);
+        if (U_FAILURE(*err)) goto exit;
+
+        // Allocate some additional strings for test values.
+        UChar testTimeZoneId[TZID_LENGTH];
+        UChar testDisplayName[DISPLAY_NAME_LENGTH];
+        UChar testDisplayName2[DISPLAY_NAME_LENGTH];
+
+        // Enumerate over all the time zones having the same base offset.
+        UEnumeration* pEnum = ucal_openTimeZoneIDEnumeration(UCAL_ZONE_TYPE_CANONICAL_LOCATION, NULL, &baseOffset, err);
+        if (U_FAILURE(*err)) goto exitEnum;
```

Comment:
>exitEnum [](start = 34, length = 8)

this will jump to place closing the enumeration while the opening operation is failed.